File: CPP/No1143.cc
/**
 * Created by Xiaozhong on 2020/10/27.
 * Copyright (c) 2020/10/27 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <string>
#include <iostream>
#include <cassert>

using namespace std;

/**
 * 动态规划，是一种双串匹配 DP 模型，在这里初始化二维的动态记忆矩阵 dp(n + 1, m + 1)
 * 然后使用状态方程进行更新，dp[i][j] 表示到串 1 的第 i 个字符和串 2 的第 j 个字符最长的子串长度
 * 1. 如果 [i - 1] == [j - 1] 那么 dp[i][j] = dp[i - 1][j - 1] + 1
 * 2. 否则 dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
 */
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int n = text1.length(), m = text2.length();
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= m; ++j) {
                if (text1[i - 1] == text2[j - 1]) dp[i][j] = dp[i - 1][j - 1] + 1;
                else dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
        return dp[n][m];
    }
};

int main() {
    Solution s;
    assert(s.longestCommonSubsequence("abcde", "ace") == 3);
}

File: Java/src/No1143.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

public class No1143 {
    public int longestCommonSubsequence(String text1, String text2) {
        int m = text1.length(), n = text2.length();
        int[][] dp = new int[m + 1][n + 1];
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                // 如果 t1[i] == t2[j] 那么就可以从 dp[i - 1][j - 1] 转移过来，加上 1 就好了
                if (text1.charAt(i) == text2.charAt(j)) dp[i + 1][j + 1] = dp[i][j] + 1;
                // 否则的话，就看看从哪个方向过来能得到最大值
                else dp[i + 1][j + 1] = Math.max(dp[i][j + 1], dp[i + 1][j]);
            }
        }
        return dp[m][n];
    }

    public static void main(String[] args) {
        No1143 no1143 = new No1143();
        System.out.println(no1143.longestCommonSubsequence("abcde", "ace"));
    }
}
